src/coprocessor/codec/mysql/duration.rs

```diff
@@ -32,13 +42,165 @@ fn check_dur(dur: &StdDuration) -> Result<()> {
+//
+// 1. The behavior of this function is similar to `read_int` except that it's designed to read the
+// fractional part of a `TIME`, it will never round with `fsp` and the round will be delayed to
+// the construction of `TIME`.
+// 2. The fractional part will be align to a 9-digit number which it's easy to round with `fsp`
+named_args!(read_int_with_fsp(fsp: u8)<u32>, map_res!(
+        digit1,
+        |buf: &[u8]| -> Result<u32> {
+            let fsp = fsp as usize;
+            let (fraction, len) = if fsp >= buf.len() {
+                (buf_to_int(buf), buf.len())
+            } else {
+                (buf_to_int(&buf[..=fsp]), fsp + 1)
+            };
+            Ok(fraction * TEN_POW[9 - len])
+        }
+));
+
+// Parse the sign of `Duration`, return true if it's negative otherwise false
+named!(
+    neg<bool>,
+    do_parse!(
+        neg: map!(opt!(complete!(char!('-'))), |flag| flag.is_some())
+            >> cond_with_error!(
+                neg,
+                alt_complete!(
+                    preceded!(multispace1, preceded!(digit0, tag!(".")))
+                        | peek!(digit1)
+                        | peek!(tag!("."))
+                )
+            )
+            >> (neg)
+    )
+);
+
+// Functionality:
+// Parse the day/block(format like `HHMMSS`) value of the `Duration`
+//
+// 1. Assuming that there is no integers before marker {integer}.
+// 2. We can't determine the value we extract is a `day` or a `block`(`HHMMSS`) before we iterate
+//    the whole string.
+// 3. If a integer exists in a string like `{integer} {some digit}`, then it should parse as a
+//    `day` value.
+// 4. If a integer exists in a string like `{integer}.`, then it should parse as a `block`
+// 5. If a integer exists in a string like `{integer}\s*$`, then it should parse as a `block` too.
+
+named!(
+    day<Option<u32>>,
+    opt!(do_parse!(
+        day: read_int
+            >> alt_complete!(
+                preceded!(multispace1, peek!(digit1))
+                    | preceded!(multispace0, peek!(tag!(".")))
+                    | preceded!(multispace0, eof!())
+            )
+            >> (day)
+    ))
+);
+
+// Parse format like: `hh:mm` `(day) hh` `hh:mm:ss`
+named!(
+    hhmmss<(Option<u32>, Option<u32>, Option<u32>)>,
+    do_parse!(
+        hour: opt!(map_res!(read_int, check_hour))
+            >> has_mintue: map!(opt!(complete!(char!(':'))), |flag| flag.is_some())
+            >> minute: cond_with_error!(has_mintue, map_res!(read_int, check_minute))
+            >> has_second: map!(opt!(complete!(char!(':'))), |flag| flag.is_some())
+            >> second: cond_with_error!(has_second, map_res!(read_int, check_second))
+            >> (hour, minute, second)
+    )
+);
+
+// Parse fractional part.
+named_args!(
+    fraction(fsp: u8)<Option<u32>>,
+    preceded!(opt!(complete!(char!('.'))), opt!(call!(read_int_with_fsp, fsp)))
+);
+
+named_args!(parse(fsp: u8)<
```

Comment:
Could we write a BNF for what we are going to parse?

If i'm correct, the gramma looks like:

1. An optional `-`
2. An HHMMSS block, or a DAY (shoud follow ` `), or an HH (should follow `:`), or fraction part (leading with `.`)
3. For HHMMSS block, there can be fraction part later.
4. For DAY block, there can be HH or fraction part later.
5. For HH, there can be MM or fraction part later.
etc

Reply:
I mean, this function should be the function that contains a lot of condition checks and dispatch to different parser combinators. Other parser combinators should be pretty simple and only do what they supposed to do.